Caliper support (#1773)
LLNL/RAJA: docs/sphinx/user_guide/profiling_with_caliper.rst

within the program.

.. note:: RAJA methods which offload to the GPU will require a synchronous GPU policy to ensure that the kernel
          has completed prior to the total runtime being reported. For asynchronous kernels Caliper offers the
          `hip.gputime` or `cuda.gputime` service which may be added to CALI_CONFIG to accurately capture kernel
          time.

========================================
Profiling RAJA kernels via kernel naming
========================================
Caliper annotations of RAJA kernels work through the RAJA kernel naming mechanism currenly only supported in forall
and launch. The ``RAJA::expt::KernelName`` container holds a string and used for profiling in Caliper. Kernels
which are not provided a name are ommited from Caliper profiling::

    RAJA::forall<RAJA::seq_exec>(RAJA::RangeSegment(0, N),
    RAJA::expt::KernelName("RAJA Seq daxpy Kernel"), [=] (int i) {

        a[i] += b[i] * c;

  });

.. note:: The RAJA KernelName feature lives under the expt namespace as it part of a new param reducer interface.
          It will be removed from expt once the new reducer interface has matured.

> .. note:: The RAJA KernelName must be placed before the lambda to ensure proper behavior.

Reviewer: Probably something to discuss elsewhere, but is KernelName really necessary or could it just be Name to be more generic? There are places where multiple kernels may be called within a construct like kernel, scan, or sort.

Author: @LLNL/raja-core any objections to changing "KernelName ->Name"?  This will break the API for forall.

Author: I'll go with ``Name`` unless someone objects...

Author: It is now name.... 👍

Author: ah that is right, I'll move it out of expt as well.

.. ##
.. ## Copyright (c) 2016-25, Lawrence Livermore National Security, LLC
.. ## and RAJA project contributors. See the RAJA/LICENSE file
.. ## for details.
.. ##
.. ## SPDX-License-Identifier: (BSD-3-Clause)
.. ##

.. _profiling-with-Caliper-label:

************************
Profiling with Caliper
************************

The RAJA abstraction layer has dynamic and static plugin support for the Caliper performance library.
We refer the reader to the :ref:`feat-plugins-label` section for more on RAJA plugins. Caliper is developed at LLNL and freely available on GitHub,
see `Caliper GitHub <https://github.com/LLNL/Caliper>`_ . Caliper provides a common interface for various vendor profiling tools and its own built-in performance
reports.

> In this section we demonstrate how to configure RAJA with Caliper, run a simple examples with kernel profiling,

Reviewer 2: ```suggestion
In this section we demonstrate how to configure RAJA with Caliper, run simple examples with kernel profiling,
```

> and a quick snapshot of hotspots of using the Thicket libary. The Thicket library is also developed at LLNL and freely available on GitHub,

Reviewer 2: ```suggestion
and view a quick snapshot of hotspots of using the Thicket libary. The Thicket library is also developed at LLNL and freely available on GitHub,
```

see `Thicket GitHub <https://github.com/LLNL/Thicket>`_ .

> This page is tailored to be a quickstart guide, for more detailed tutorials we refer the reader to the Caliper and Thicket tutorials.

Reviewer 2: Would it be useful to add links to those tutorials or note where they can be found (e.g., linked to Caliper and Thicket GitHub project pages)?

Author: Those all clickable links that will take the user directly to their page.

Reviewer 2: What I meant was change "we refer the reader to the Caliper and Thicket tutorials" to "we refer the reader to the Caliper and Thicket tutorials, which are accessible on the GitHub pages for those projects"

=================================
Building and running with Caliper
=================================

> Caliper serves as a portable profiling library which may be configured with various vendor options. For the most up to date

Reviewer 2: Please add a blank line before section header and text.

the following runtime information::

  Path                     Time (E) Time (I) Time % (E) Time % (I)
  C-version elapsed time   0.000820 0.000820   0.395169   0.395169
  RAJA Seq daxpy Kernel    0.000655 0.000655   0.315502   0.315502
  RAJA SIMD daxpy Kernel   0.000611 0.000611   0.294629   0.294629
  RAJA OpenMP daxpy Kernel 0.013691 0.013691   6.598422   6.598422
  RAJA CUDA daxpy Kernel   0.000118 0.000118   0.056827   0.056827

The (E) column corresponds to exclusive timing; while the (I) column will correspond to inclusive timing.
The left two columns are absolute time in seconds while the right two columns correspond to percentage of time
within the program.

.. note:: RAJA methods which offload to the GPU will require a synchronous GPU policy to ensure that the kernel
          has completed prior to the total runtime being reported. For asynchronous kernels Caliper offers the
          `hip.gputime` or `cuda.gputime` service which may be added to CALI_CONFIG to accurately capture kernel
          time.

========================================
Profiling RAJA kernels via kernel naming
========================================

> Caliper annotations of RAJA kernels work through the RAJA kernel naming mechanism which is currently only supported in forall

Reviewer 2: Please add a blank line before section header and text.

> and launch. The ``RAJA::Name`` container holds a string and used for profiling in Caliper. Kernels

Reviewer 2: ```suggestion
and launch. The ``RAJA::Name`` container holds a string and is used for profiling in Caliper. Kernels
```

> which are not provided a name are ommited from Caliper profiling::

Reviewer 2: I think this sentence should come after the following code example which shows a kernel with a name.

    RAJA::forall<RAJA::seq_exec>(RAJA::RangeSegment(0, N),
    RAJA::Name("RAJA Seq daxpy Kernel"), [=] (int i) {

        a[i] += b[i] * c;

  });

.. note:: The RAJA Name feature lives under the RAJA namespace.

.. note:: The RAJA Name may be placed anywhere in the RAJA reducer parameter list

.. note:: When RAJA is not configured with Caliper, NVTX and ROCTX profiling can
          still be configured with RAJA using the following CMAKE variables respectively::
            -DRAJA_ENABLE_NV_TOOLS_EXT=ON
            -DRAJA_ENABLE_ROCTX=ON

>          When Caliper is configured with RAJA all performance profiling is performed by Caliper.

Reviewer 2: Is this last sentence also supposed to be a note? If so, it is missing the note annotation. If not, it should be separated from the previous note by a blank line and not indented.

=============================================
Basic integration with vendor profiling tools
=============================================

> Once Caliper is configured with RAJA and kernels are provided with a kernel name, the Caliper library provides various

Reviewer 2: Please add a blank line before section header and text.

> Once Caliper is configured with RAJA and kernels are provided with a kernel name, the Caliper library provides various

Reviewer 2: ```suggestion
Once RAJA is configured with Caliper and kernels are provided with kernel names, the Caliper library provides various
```

Also, please add a blank line before section header and text.

  RAJA CUDA daxpy Kernel
  |-                                                                         0.000159
  |-                      void RAJA::policy::cuda~~}::detail::Name>)            0.000038
  cudaLaunchKernel
   |-                                                                        0.000066
   |-                    void RAJA::policy::cuda~~}::detail::Name>)            0.000038
  cudaStreamSynchronize                                                      0.000050
  cudaFree                                                                   0.000495

A simiar command may be found for HIP kernels under the Caliper documentation.

^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Generating an NVIDIA Nsight Systems file
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

Caliper can also be used to generate a NVIDIA Nsight Systems file. The following command will generate the file and annotate the CUDA
kernel region::

  CALI_SERVICES_ENABLE=nvtx,cuda nsys profile -o my_profile ./bin/raja-forall-caliper

The nsys file may then be read into the NVIDIA Nsight toolkit.

.. image:: figures/CUDA_profiling.png

^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Generating a ROCM trace file
^^^^^^^^^^^^^^^^^^^^^^^^^^^^

> Configuring a ROCM trace file with Caliper may be done using the following command::

Reviewer 2: Please add a blank line before section header and text.

  CALI_SERVICES_ENABLE=roctx,rocm,trace rocprof --hip-trace --roctx-trace ./bin/raja-forall-caliper

The trace file may then be read into the Chrome tracer tool as demonstrated below.

.. image:: figures/ROCM_profiling.png

^^^^^^^^^^^^^^^^^^^^^^^
Generating Hatchet file
^^^^^^^^^^^^^^^^^^^^^^^

Lastly, the following command will generate a .cali file which can be read into the Thicket library::

  CALI_CONFIG=hatchet-region-profile ./bin/raja-forall-caliper

> Using python jupyter notebooks and the Hatchet library we can load the .cali file and begin analysing

Reviewer 2: ```suggestion
Using python jupyter notebooks and the Hatchet library we can load the .cali file and begin analyzing
```

performance of our application. The following sequence of commands will generate a color coded tree
high-lighting hotspots (in red) of our application::

  caliper_file = region_profile.cali
  gf = ht.GraphFrame.from_caliperreader(caliper_file)
  print(gf.tree())

Within the jupyter notebook the color coded tree will look like this:

.. image:: figures/Hatchet_tree.png

=============
Final remarks
=============

> The capabilities of Caliper and Hatchet exceed what has been presented here. Our main goal was to provide

Reviewer 2: Please add a blank line before section header and text.
